fpga_interchange/static_timing_analysis.py

```diff
@@ -0,0 +1,583 @@
+                            temp.remove((temp_obj.site, temp_obj.bel,
+                                         temp_obj.pin))
+
+                    if self.detail:
+                        indent += 1
+                        if len(temp) > 0:
+                            print("\t" * indent + "Exploring",
+                                  self.phy_netlist.strList[obj.site],
+                                  self.phy_netlist.strList[obj.bel],
+                                  self.phy_netlist.strList[obj.pin],
+                                  "found bels:")
+                        indent += 1
+                        for new_end in temp:
+                            print("\t" * indent,
+                                  self.phy_netlist.strList[new_end[0]],
+                                  self.phy_netlist.strList[new_end[1]],
+                                  self.phy_netlist.strList[new_end[2]])
+                        indent -= 2
+                    old_branches = vertex.disown('branches')
+                    vertex.init('branches',
+                                len(old_branches.get()) + len(temp))
+                    for i, branch in enumerate(old_branches.get()):
+                        vertex.branches[i] = branch
+                    for i, new_end in enumerate(temp):
+                        branch = vertex.branches[len(old_branches.get()) + i]
+                        branch.routeSegment.init('belPin')
+                        branch.routeSegment.belPin.site = new_end[0]
+                        branch.routeSegment.belPin.bel = new_end[1]
+                        branch.routeSegment.belPin.pin = new_end[2]
+            elif which == "sitePin":
+                obj = vertex.routeSegment.sitePin
+            elif which == "pip":
+                obj = vertex.routeSegment.pip
+            elif which == "sitePIP":
+                obj = vertex.routeSegment.sitePIP
+            last = len(vertex.branches) == 0
+            if not last:
+                for branch in vertex.branches:
+                    dfs_traverse(branch, False)
+            elif not start:
+                assert which == "belPin", obj
+                ends_array.append((vertex, (obj.site, obj.bel, obj.pin)))
+            return
+
+        if self.detail:
+            indent += 1
+            print("\t" * indent + f"{self.phy_netlist.strList[net.name]}")
+        for i, source in enumerate(net.sources):
+            ends_array = []
+            which = source.routeSegment.which()
+            if which == "belPin":
+                obj = source.routeSegment.belPin
+                sources_array.append((i, (obj.site, obj.bel, obj.pin)))
+                dfs_traverse(source, True)
+            else:
+                raise
+            sinks_array.extend(ends_array)
+        if self.detail:
+            print("\t" * indent + "Searching for pseudo sitePIPs")
+        # assumption is that if some bel has both net sink and source it's probably pseudo sitePIP
```

Comment:
just to note that flipflops or BRAMs with feedback to themselves aren't totally impossible, although it's true this is the most likely case